draft-irtf-cfrg-opaque.md

```diff
@@ -2127,10 +2126,11 @@ disclose their passwords to the server, even during registration. Note that a co
 server can run an exhaustive offline dictionary attack to validate guesses for the client's
 password; this is inevitable in any aPAKE protocol. (OPAQUE enables defense against such
```

Comment:
Since threshold OPAQUE is out of scope and, as far as I know, would require protocol changes, I would just remove this entire sentence ("(OPAQUE enables ... compromised {{JKX18}}.)") from the document.

Reply:
that depends on how you define protocol changes, yes a parameter to oprf_finalize would be larger (since it contains the blinded/evaluated shares that need to be combined before unblinding) but besides that the only thing that needs change is oprf_evaluate which acts as a proxy to the shareholders. but yeah that still doesn't put threshold in scope of this spec. which is a pity though (same for the oprf spec which lacks threshold variant, and would make this whole thing easier also in opaque). 

anyway i would leave mentioning this mitigation in here but declare it out of scope for this spec.

Reply:
Minimally, clients would need to be aware of the different threshold servers and then generate messages for each. Right now, there is exactly one server in the protocol, and doing anything different seems like it would pretty clearly require a change.

Reply:
oh, actually! no, it is possible to do everything in oprf_evaluate! and nothing in the oprf_evaluate fn params would change, oprf_evaluate would simply act as a proxy that distributes the evaluation among the shareholders and then combines the result and returns that as the result of oprf_evaluate. from the opaque perspective this would be completely "opaque" and noone would have any idea if the oprf output is a threshold or non-threshold version. yes, of course oprf_evaluate would need to know about the shareholders and the threshold. but that is an oprf_evaluate internal setting and can be totally ignored by opaque.

Reply:
i will implement this asap!

Reply:
That sort of deployment is indeed possible, but I don't think it faithfully captures the intent of the paper here. The proxy server can still be compromised, such as a single server could, to make everything fall apart. I claim that "true" benefits to adding threshold support only come if the client is actually a participant in that protocol.

Reply:
and i'm sure i will still satisfy the testvectors and will not have to change anything except for oprf_evaluate!

Reply:
@stef see above -- I suspect we have different mental models for how one might use a threshold OPRF here.

Reply:
> That sort of deployment is indeed possible, but I don't think it faithfully captures the intent of the paper here. The proxy server can still be compromised, such as a single server could, to make everything fall apart. I claim that "true" benefits to adding threshold support only come if the client is actually a participant in that protocol.

hmmm right, if the opaque server and the threshold-oprf shareholder servers are all under the control of the attacker then this:

> server can run an exhaustive offline dictionary attack to validate guesses for the client's
password

will still work. but if the oprf shareholders are not under the attackers control, then it is a mitigation.

but then the other solution where the combination of the shares happens in oprf_finalize also makes no difference even though it happens at the client, since that can be also faked by the offline bruteforcing server.

Reply:
ok, ok, i'm holding my horses. but now i wanna know how to do this to claim the true benefits of a toprf. @hugokraw can you enlighten us please?

Reply:
hmmm, actually reading the paper:

> Threshold implementation. We comment on a simple extension of OPAQUE
> that can be very valuable in large deployments, namely, the ability to implement
> the OPRF phase as a Threshold OPRF [28]. In this case, an attacker needs to
> break into a threshold of servers to be able to impersonate the servers to the user
> or to run an offline dictionary attack. Such an implementation requires no user-
> side changes, i.e., the user does not need to know if the system is implemented
> with one or multiple servers.

this refers to:
S. Jarecki, A. Kiayias, H. Krawczyk, and J. Xu. TOPPSS: Cost-minimal password-protected secret sharing  based on threshold OPRF. In Applied Cryptology and Network Security – ACNS 2017, pages 39–58. Springer, 2017.

which says this:

> This contribution is based on the observation that a more efficient PPSS can result from replacing the  OPRF used in the protocols of [18,19] with its threshold (or multi-party) counterpart which we define as Threshold OPRF (T-OPRF).

Reply:
i just figured if we don't trust the opaque server then a dkg is necessary.

Reply:
@chris-wood I understand your point, but I advocate for still mentioning it since the paper talks about it.

The current phrase can be misleading, so I suggest adding to `10.1. Notable Design Differences` that this spec does not consider thresholding. What do you think?

Reply:
I'd be OK mentioning it in the notable design differences section.

Reply:
@bytemare will you be able to update this text?

Reply:
Please let me know what you think :)